### PR TITLE
Replace all formatters and linters with Ruff

### PR DESCRIPTION
**Pull request by wilbrdt**
## Purpose

Ruff is extremely fast and popular.
Replacing our tools with it for linting. We're for now keeping black for formatting, both locally and in CI.

## Proposal

It removes all linter tools (but black) that are:
- isort
- flake8
- pylint
- bandit
- pydocstyle 

and replace them with Ruff.

Note: gitlint will not pass valid for this PR as `.pylintrc` file is being deleted, and contains the word `FIXME`
Commits in this PR have a high granularity to (hopefully) ease the reviewability.
But they should be squashed (into probably 3 or 4 commits max) before merging.

### CODE REVIEW

File: tests/backends/data/test_fs.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 def test_backends_data_fs_data_backend_default_instantiation(monkeypatch, fs):
     """Test the `FSDataBackend` default instantiation."""
-    # pylint: disable=invalid-name
```
Comment (Leobouloc) [inline tests/backends/data/test_fs.py:19]: don't we want this checked ? (invalid-name)

File: tests/backends/data/test_fs.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def test_backends_data_fs_data_backend_instantiation_with_settings(fs):
 
     try:
         FSDataBackend(settings)
-    except Exception as err:  # pylint:disable=broad-except
```
Comment (Leobouloc) [inline tests/backends/data/test_fs.py:67]: don't we want this checked ?

File: tests/backends/data/test_ldp.py [hunk 3]
```diff
@@ -59,7 +59,7 @@ def test_backends_data_ldp_data_backend_instantiation_with_settings(ldp_backend)
 
     try:
         ldp_backend(service_name="bar")
-    except Exception as err:  # pylint:disable=broad-except
```
Comment (Leobouloc) [inline tests/backends/data/test_ldp.py:62]: Shouldn't we keep this ?

File: tests/backends/data/test_mongo.py [hunk 4]
```diff
@@ -852,7 +852,7 @@ def test_backends_data_mongo_data_backend_write_method_with_custom_chunk_size(
     """Test the `MongoDataBackend.write` method, given a custom chunk_size, should
     insert the provided documents to target collection by batches of size `chunk_size`.
     """
-    # pylint: disable=unused-argument
```
Comment (Leobouloc) [inline tests/backends/data/test_mongo.py:855]: shouldn't we keep this ?

**Review by Leobouloc — COMMENTED** [review]
Looking forward to working with this !

My main questions are concerning checks which seem to longer by made ("invalid-name", "unused-arguments", etc).

File: tests/backends/data/test_ldp.py (hunk 3 above)
Comment (Leobouloc) [reply tests/backends/data/test_ldp.py:62]: same question as previous

File: docs/contribute.md [hunk 5]
```diff
@@ -38,8 +38,8 @@ To lint your code, either use the `lint` meta target or one of the linting tools
 # Run all linters
 $ make lint
 
-# Run pylint
-$ make lint-pylint
+# Run ruff linter
+$ make lint-ruff
```
[the remark this answers is not in the document]
Comment (wilbrdt) [reply docs/contribute.md:42]: For sure!

File: pyproject.toml [hunk 6]
```diff
@@ -150,30 +144,62 @@ universal = true
...
+]
+
+[tool.ruff.lint.isort]
+section-order = ["future", "standard-library", "third-party", "ralph", "first-party", "local-folder"]
+[tool.ruff.lint.isort.sections]
+ralph = ["ralph"]
+
+[tool.ruff.per-file-ignores]
+"tests/*" = [
+    "D", # pydocstyle
```
[the remark this answers is not in the document]
Comment (wilbrdt) [reply pyproject.toml:192]: Previously, we were also ignoring `pydocstyle` checks on tests (with the line `match_dir = "^(?!tests|venv|build|scripts).*"`).
Do you want to lint docstrings in tests? :grin:
I'm fine without it!

File: src/ralph/backends/data/async_es.py [hunk 7]
```diff
@@ -112,7 +110,7 @@ async def list(
             yield index
 
     @async_enforce_query_checks
-    async def read(
+    async def read(  # noqa: PLR0912, PLR0913
```
[the remark this answers is not in the document]
Comment (wilbrdt) [reply src/ralph/backends/data/async_es.py:113]: The default number of branches is 12 for both pylint and Ruff. Ruff counts 13 branches on this function.
I guess they don't count in the same way

File: tests/backends/data/test_es.py [hunk 8]
```diff
@@ -770,7 +761,7 @@ def test_backends_data_es_data_backend_close_method(es_backend, caplog):
 
     # No client instantiated
     backend = es_backend()
-    backend._client = None  # pylint: disable=protected-access
+    backend._client = None
```
[the remark this answers is not in the document]
Comment (wilbrdt) [reply tests/backends/data/test_es.py:764]: You're right I should have dug more into it!
`protect-access` seems to be covered by SFL001 `private-member-access`, adding it

File: tests/backends/data/test_fs.py (hunk 1 above)
Comment (wilbrdt) [reply tests/backends/data/test_fs.py:19]: You're right!
Adding the `N` set of rules (for `pep8-naming`) but escaping some of them (`N806`, `N815` and `N818`) that could be potentially escaped or corrected in the future.
Note that it does not enforce a minimum length for variables (I fiund that a minimum of 3 is a big constraint), so it won't trigger a warning on `fs` here.

File: tests/backends/data/test_fs.py (hunk 2 above)
Comment (wilbrdt) [reply tests/backends/data/test_fs.py:67]: Unfortunately, it seems there is no equivalent rule for raising a `too broad except` in Ruff.
There is the `TRY` set of rules for `tryceratops` but it doesn't raise a warning for this, and we're far from following the good exception handling patterns (lots of warning raised) :grin:

File: tests/backends/data/test_mongo.py (hunk 4 above)
Comment (wilbrdt) [reply tests/backends/data/test_mongo.py:855]: You're right!
Adding `ARG` set of rules (`flake8-unused-arguments`) that should raise warnings for this.

**Review by quitterie-lcs — APPROVED** [review]

File: pyproject.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Leobouloc) [reply pyproject.toml:192]: I would slightly lean towards including them but don't have a strong opinion.

File: tests/backends/data/test_es.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (wilbrdt) [reply tests/backends/data/test_es.py:764]: After discussion, adding `SFL` set of rules!

File: tests/backends/data/test_fs.py (hunk 1 above)
Comment (wilbrdt) [reply tests/backends/data/test_fs.py:19]: Dropping this as it is too restrictive

File: tests/backends/data/test_fs.py (hunk 2 above)
Comment (wilbrdt) [reply tests/backends/data/test_fs.py:67]: Adding `BLE` set of rules

File: tests/backends/data/test_mongo.py (hunk 4 above)
Comment (wilbrdt) [reply tests/backends/data/test_mongo.py:855]: After discussion, adding `ARG` set of rules but ignoring the tests files (it is really a pain for fixtures)

File: pyproject.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (wilbrdt) [reply pyproject.toml]: Sticking to not linting docstring in tests!

**Pull request merged**